### ISSUE TITLE
Couple of questions

### ISSUE THREAD

**Issue by PavelVozenilek** [jfecher/ante#12]
1. Will Ante support namespaces? If you are interested I have idea how to make them better than e.g. in C++.

2. String interpolation: will it be possible to "intercept" this mechanism for certain specific cases? What I have in mind is advanced assert:

  ```
    assert(x > 0, "x = ${x}, y = ${y}, z = ${z}")
  ```

  If assert happens I would like to use my own memory handling management when showing all these details, w/o worrying about accidentally running out of memory.

**Comment by jfecher** [comment]
1.  Namespaces in Ante are currently made through type extensions as the type does not have to exist for the extension to be made so, you could eg. make a function Foo.bar which would essentially be in the Foo namespace.  I'm interested to see your ideas however.

2. String interpolation always expands into the ++ append operator.  Your example expands into 
```
assert (x > 0) ("x = " ++ x ++ ", y = " ++ y ++ ", z = " ++ z)
```
So if you want to do your own memory alllocation, just overload the ++ operator for the types of x y and z.  If memory was a concern you could eg. make it so that when ++ is called it returns a lazy stream or something.

**Comment by PavelVozenilek** [comment]
Ad 1. My idea was to avoid things like:

```
namespace xyz {
...
... thousand of lines
...
}
```

because they are not always needed but fixed in stone. This syntax is also hard to follow (no one would write a function spanning thousand lines, for namespace it is normal).

Instead, a module would be created w/o any thinking about namespaces, just with short names, as if this is a small project that doesn't need this abstraction. Such code would be easier to grok

If, however, project gets big and there are name clashes, then the _user_ may add namespace according to his taste:

```
import "foo" as bar
import some_long_name as q

bar.baz()
baz() // if there are no clashes it is the same as bar.baz()
var q.some_type x = 1
```

The main advantage is that library writers won't be encouraged to design namespace hierarchies just because it is cool and _may_ be useful someday. User code which does not need namespaces won't be pessimized.

-----------------------

There's yet one thing that may be handy: namespaces as suffix. Instead of

 `std.something()`
or 
  `my_cool_namespace_for_projext_xyz_version_2_7_96.x = 1`

one may prefer writing it the other way:

   `insert/queue_lib()`
or
  `fopen{std}()`

Here, the most important part of the name is the _first one_ to read, not the administrative noise. 

The import syntax could be e.g.

  `import foo as /bar`
or 
  `import foo as {bar}`

--------------------------------

Ad 2: what I had in mind was to (somehow) use custom memory management for single special case, the `assert`. W/o any change for the rest of the code.  Unless the user decides to read `assert` details he wouldn't notice anything. 

I imagined it as some evil (but otherwise invisible) trickery with the AST.

**Comment by jfecher** [comment]
1. Your first section with namespaces is pretty much the form Ante currently uses (although admittedly it isnt in language.an).  I'm not sure about namespace as suffixes though.  I think I would rather keep the current form.

2. Yeah, if you wanted to do some memory trickery with `assert` specifically you would have two options:

    1. Accept a stream parameter / parameter of your choice and do the memory allocation trickery within the function itself.  This imo would probably be the better choice.  You could use a type of your own design to do the allocation and either require it as a parameter or rely on an implicit conversion and have users enter string parameters or something else to be converted.

    2.  Make the function a ![macro] function and do some AST trickery like you said.  This is necessary because Ante uses eager evaluation, so the parameters are compiled before the function is called.

**Comment by PavelVozenilek** [comment]
In my experience namespaces are prone to misuse, to do things "that may be useful some day".

Two examples:

1. Java with their `COM.sun.library.blah.bleh` syntax. This got aped to the extreme, for projects where it made no sense, just because of the "Java way" feeling.

2. Modern C++ with `namespace { }` replacing the old `static` label next to the function. The simple and easy way to annotate important attribute of the function is out, the tricky and hard to spot mechanism is in.

Perhaps namespaces could be dropped and replaced by mandatory renaming for clashes, like:

```
import foo {
   old_name -> new_name
   x -> foo.x // the dot has no semantic meaning here
}
```

**Comment by jfecher** [comment]
Namespaces in ante are determined by the file name.  There is no requirement to specify a `module MyModName` at the top of a file (unless you want to rename it) to prevent naming clashes because if a clash occurs the compiler issues an error even if the functions are mangled with different parameters.  Name clashes are manually resolved by either doing an `import Foo as Bar` or by excluding the conflicting function(s) `import Foo except baz`

**Comment by PavelVozenilek** [comment]
> if a clash occurs the compiler issues an error even if the functions are mangled with different parameters

This is perhaps too strict and may push people to making long names, just in case.

I know the language only from `language.an`, so I am in fog about features not mentioned there. (I expect some day there will be Windows version to play with.)

--------------
Here's one more idea, related to modules.

Say there's file `my_interface.an`. If directory `my_interface` is in the same place then all files inside this directory would be treated as implementation of the interface. This implementation would be completely internal and isolated. Nothing would escape out (no matter of access attributes or namespaces), nobody (but `my_interface.an`) would be able to import anything from this directory. 

This feature would allow one to take any 3rd party library and wrap it w/o fears of being accidentally misused later. It may also allow to have separate variants of the same code in the library, w/o risk of name clashes.

**Comment by jfecher** [comment]
> This is perhaps too strict and may push people to making long names, just in case.
I do not think so.  It is exactly how D does it, and I feel that it is beneficial to warn the user of this newly imported function that may be separate from the other implementation.  Instead of changing library names, programmers are encouraged to just keep names short and if an error does occur, simply change the `import Foo` to `import Foo as Bar` or something.

**Comment by PavelVozenilek** [comment]
It is only when D compiler sees valid overloads from different modules then it shows error, instead of picking up the most fitting one. 

http://dlang.org/spec/function.html#overload-sets

Like:

```
--- module A ---
foo(int a) ...
--- module B ---
foo(long a) ...
---- main file --
import A, B
foo(1) // error, two options from two modules
```


IOW, overloading by parameter types is allowed as long as it is not likely result of programmer's mistake.

**Comment by jfecher** [comment]
Ok there was just some confuaion on my part when the error takes place then.  So Ante will be like D in this case as well.
